In [4]:
import os
os.environ['CUDA_VISIBLE_DEVICES'] = "4,6,7"
import torch
import torch.nn as nn
import torch.optim as optim
from transformers import AutoModel, AutoTokenizer, pipeline, AutoModelForCausalLM
from swarm.optimizer.edge_optimizer.edge_network import EdgeNetwork
from swarm.optimizer.edge_optimizer.parameterization import EdgeWiseDistributionByModel
import pickle as pkl
import numpy as np
import re

from swarm.graph.swarm import Swarm
from typing import List, Any, Optional

from swarm.llm.custom_llm import CustomLLM
from swarm.llm.format import Message

from vllm import LLM, SamplingParams
import matplotlib.pyplot as plt

from experiments.evaluator.datasets.mmlu_dataset import MMLUDataset
from experiments.evaluator.datasets.cmmlu_dataset import CMMLUDataset
from experiments.evaluator.datasets.mixedmmlu_dataset import MixedMMLUDataset
import pickle
import json
import asyncio
import json
from huggingface_hub import login 
login("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")

Token has not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.
Token is valid (permission: write).
Your token has been saved to /home/lucas/.cache/huggingface/token
Login successful


In [5]:
import pickle
import io

class CPU_Unpickler(pickle.Unpickler):
    def find_class(self, module, name):
        if module == 'torch.storage' and name == '_load_from_bytes':
            return lambda b: torch.load(io.BytesIO(b), map_location='cuda:7')
        else:
            return super().find_class(module, name)


In [6]:
path = "mixedmmlu_8true_0adv_edge_iter200_domain_mixedmmlu_MajorityVote_OptimizedSwarm_1714788267.6306956" + ".pkl"
with open(f"result/mmlu/{path}", 'rb') as f:
    swarm = pickle.load(f)  # Load the swarm
    #swarm = CPU_Unpickler(f).load()
    #delete swarm object and clear memory
    score = pickle.load(f)  # Load the score
print(path)
print(score)

mixedmmlu_8true_0adv_edge_iter200_domain_mixedmmlu_MajorityVote_OptimizedSwarm_1714788267.6306956.pkl
0.524


In [27]:
for i,node in enumerate(swarm.composite_graph.nodes.values()):
    if i ==0:
        continue
    print(node.llm.models)

{}
{}
{}
{}
{}
{}
{}
{}


In [2]:

with open(f"result/crosswords/old_method_10_it_nostuck_20batch/experiment8_edge_logits_9.pt", 'rb') as f:
    edge_logits = torch.load(f)  # Load the swarm
    #delete swarm object and clear memory
print(edge_logits)

OrderedDict([('edge_logits', tensor([-0.9060, -2.0771, -0.3538, -1.9184, -2.3243, -3.0701, -3.8445, -3.8105,
        -2.0401, -1.5536, -2.3212, -4.7880, -1.9222, -1.7154, -4.7214, -2.5089,
        -2.1340,  0.9295, -2.5914,  1.2137, -2.2820, -2.1987, -3.1128,  0.2236,
        -1.2063, -2.0589, -4.0054, -1.2779, -2.5772, -4.5234, -2.2227, -1.8173,
        -3.7507, -2.4876, -0.0229, -2.1858])), ('order_params', tensor([ 0.2713, -1.2729,  0.5027,  0.4181, -0.6394, -0.6608, -0.1433, -0.1043,
        -1.5313]))])


In [7]:
swarm = Swarm(["CrosswordsBruteForceOpt","CrosswordsReflection"], "crosswords", "meta-llama/Meta-Llama-3-8B-Instruct",#"google/gemma-7B-it",#,#"gpt-3.5-turbo-1106", #"gpt-4-1106-preview" ,  #"CrosswordsToT","CrosswordsBruteForceOpt","CrosswordsReflection"
            final_node_class="ReturnAll", 
            final_node_kwargs={},
            edge_optimize=True,
            init_connection_probability=0.1, 
            connect_output_nodes_to_final_node=True, 
            include_inner_agent_connections=True,
            edge_network_enable=False,
            llm_backbone_name="")
swarm.connection_dist.edge_logits = nn.Parameter(edge_logits['edge_logits'])


We are using custom LLM class, model_name: meta-llama/Meta-Llama-3-8B-Instruct
Load Model...
Selected GPU: 7


Loading checkpoint shards: 100%|██████████| 4/4 [00:10<00:00,  2.64s/it]


RuntimeError: CUDA error: invalid device ordinal
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.


In [4]:
graph, _ = swarm.connection_dist.realize(swarm.composite_graph)

In [5]:
print(graph)

In [6]:
from swarm.graph.visualize import GPTSwarmVis

In [9]:

GPTSwarmVis(graph, file_name = "result.html")

/home/lucas/DynamicGPTSwarm/result/example.html


In [ ]:

from swarm.graph.swarm import Swarm
from swarm.environment.operations.final_decision import MergingStrategy
from experiments.evaluator.evaluator import Evaluator
from experiments.evaluator.datasets.mmlu_dataset import MMLUDataset
from dataset.MMLU.download import download
from experiments.evaluator.datasets.cmmlu_dataset import CMMLUDataset
from experiments.evaluator.datasets.mixedmmlu_dataset import MixedMMLUDataset
from dataset.CMMLU.download import download as cmmlu_download

from swarm.llm.custom_llm import CustomLLM

swarm = Swarm(
    ["IO"],
    "mmlu",
    model_name="meta-llama/Meta-Llama-3-8B-Instruct",
    final_node_class="FinalDecision",
    final_node_kwargs=dict(strategy=MergingStrategy.MajorityVote),
    edge_optimize=True,
    edge_network_enable=False,
    llm_backbone_name="google/gemma-2B",
)
print(swarm.connection_dist.edge_logits.shape)
#got the swarm, now evaluator
dataset_train = MMLUDataset('dev', categories=["college_mathematics","elementary_mathematics","formal_logic", "abstract_algebra", "high_school_mathematics"])
dataset_val = MMLUDataset('val',categories=["college_mathematics","elementary_mathematics","formal_logic", "abstract_algebra", "high_school_mathematics"])
evaluator = Evaluator(
    swarm,
    dataset_val,
    dataset_val,
    model_name="meta-llama/Meta-Llama-3-8B-Instruct",
    enable_tensorboard = False,
    enable_artifacts=True,
    tensorboard_tag=None)
accs = {}
accs['IO'] = await evaluator.evaluate_agent(agent="IO", limit_questions=25)
accs['COT'] = await evaluator.evaluate_agent(agent="COT", limit_questions=25)


In [ ]:
print(accs)

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

data = [0.1, 0.3, 0.1, 0.1, 0.1, 0.2, 0.2, 0.3, 0.1, 0.4, 0.1, 0.2, 0.2, 0.3, 0.1, 0.0, 0.1, 0.0, 0.1, 0.1, 0.1, 0.1, 0.2, 0.1, 0.0, 0.4, 0.2, 0.1, 0.1, 0.2, 0.4, 0.2, 0.2, 0.1, 0.2, 0.2, 0.1, 0.2, 0.2, 0.2, 0.2, 0.2, 0.1, 0.1, 0.0, 0.4, 0.2, 0.1, 0.0, 0.1, 0.1, 0.2, 0.3, 0.4, 0.5, 0.1, 0.1, 0.2, 0.1, 0.1, 0.1, 0.1, 0.0, 0.3, 0.1, 0.4, 0.1, 0.1, 0.2, 0.0, 0.0, 0.3, 0.2, 0.3, 0.1, 0.2, 0.4, 0.1, 0.2, 0.0, 0.2, 0.2, 0.0, 0.2, 0.1]

data = [[i for i in data[j:j+20]]for j in range(len(data)//20)]

print(data)


In [ ]:
lin = torch.nn.Linear(2048,36)
path = "result/crosswords/experiment8_edge_logits_0.pt"
dict = torch.load(path)
for key in dict.keys():
    print(key)
    print(dict[key].shape)
    print(dict[key])
    print("----")
lin.load_state_dict(torch.load(path))

In [ ]:
with open('result/crosswords/experiment9_edge_network-False_final_utilities.pkl', 'rb') as f:
    utilities = pkl.load(f)
print(utilities.keys())
for key in utilities.keys():
    if key !="utilities":
        print(key, " ", utilities[key])
with open('result/crosswords/experiment9_edge_network-True_final_utilities.pkl', 'rb') as f:
    utilities = pkl.load(f)
print(utilities.keys())
for key in utilities.keys():
    if key !="utilities":
        print(key, " ", utilities[key])

In [ ]:
with open('result/crosswords/experiment9_edge_network-False_final_utilities.pkl', 'rb') as f:
    utilities = pkl.load(f)
print(utilities.keys())
for key in utilities.keys():
    if key !="utilities":
        print(key, " ", utilities[key])

#turn every mean into max 3 decimals after comma

utilities['mean_utilities_per_run'] = [round(i,3) for i in utilities['mean_utilities_per_run']]
utilities['overall_mean'] = round(np.mean(utilities['mean_utilities_per_run'], axis=0),3)
utilities['overall_std'] = round(np.std(utilities['mean_utilities_per_run'], axis=0),3)
print(utilities['overall_mean'])
print(utilities['overall_std'])
print(utilities['mean_utilities_per_run'])
#dump results
with open('result/crosswords/experiment9_edge_network-False_final_utilities.pkl', 'wb') as f:
    pkl.dump(utilities, f)

In [ ]:
evaluator = CrosswordsEvaluator(test_data, batch_size=batch_size, metric="words", window_size=num_batches)
swarm = Swarm(["CrosswordsBruteForceOpt","CrosswordsReflection"], "crosswords", "meta-llama/Meta-Llama-3-8B-Instruct",#"google/gemma-7B-it",#,#"gpt-3.5-turbo-1106", #"gpt-4-1106-preview" ,  #"CrosswordsToT","CrosswordsBruteForceOpt","CrosswordsReflection"
            final_node_class="ReturnAll", 
            final_node_kwargs={},
            edge_optimize=True,
            init_connection_probability=init_connection_probability, 
            connect_output_nodes_to_final_node=True, 
            include_inner_agent_connections=True,
            edge_network_enable=edge_network_enable,
            llm_backbone_name=llm_backbone_name)

In [ ]:
def process_messages(messages: List[Message]) -> List[Message]:
        processed_messages = []
        system_message = None

        for message in messages:
            if message.role == 'system':
                system_message = message.content
            elif message.role == 'user':
                if system_message:
                    message.content = system_message + ' ' + message.content
                    system_message = None
                processed_messages.append(message)

        # Handle the case where the last message is a system message
        if system_message:
            processed_messages.append(Message(role='user', content=system_message))

        return processed_messages


In [ ]:
#load gemma-7B-it with Custom LLM class

from swarm.environment.prompt.prompt_set_registry import PromptSetRegistry
from tqdm import tqdm

model = CustomLLM("google/gemma-7B-it" )
test_dataset = CMMLUDataset("test")
prompt_set = PromptSetRegistry.get("cmmlu")

#test_dataset._total_df = test_dataset._total_df[:20]

tqdm = tqdm(test_dataset)


y = []

for sample in tqdm:
    input = test_dataset.record_to_swarm_input(sample)
    role = prompt_set.get_role()
    constraint = prompt_set.get_constraint()
    message = [Message(role="system", content=f"You are a {role}. {constraint}"),Message(role="user", content=input["task"])]

    input_dict = process_messages(message)
    answer = model.gen(message)
    answer = test_dataset.postprocess_answer(answer)
    correct_answer = test_dataset.record_to_target_answer(sample)
    y.append(answer == correct_answer)
print("final accuracy: ", np.mean(y))


In [ ]:
def apply_chat_template(messages, add_generation_prompt=True):
    if add_generation_prompt:
        if messages[-1]["role"] == "user":
            messages.append({"role": "assistant", "content": ""})
    chat = ""
    for message in messages:
        print(message)
        role, content = message.values()
        print(role, content)
        if role == "user":
            chat += f"[|Human|]:{content}"
        elif role == "assistant":
            chat += f"[|AI|]:{content}"
        else:
            raise ValueError(f"Unknown role {role}")
    return chat
tokenizer.apply_chat_template = apply_chat_template

In [ ]:
with open("result/crosswords/old_method_10_it_nostuck_20batch/experiment8_utilities_9.pkl", "rb") as f:
    utilities = pkl.load(f)
#print(utilities[100:])
print(np.mean(utilities[100:]))
for i in range(0,200,20):
    print(utilities[i:i+20])
    print(np.mean(utilities[i:i+20]))

In [ ]:
print(torch.sigmoid(swarm['edge_logits']))


In [5]:
def create_adjacency_matrix(edge_probs, swarm):
    assert swarm is not None

    # Create an empty adjacency matrix
    num_nodes = len(swarm.composite_graph.nodes)
    adjacency_matrix = np.zeros((num_nodes, num_nodes))

    # Create a dictionary to map indices to nodes
    node_indices = {}
    node_names = {}
    for index, (node_id, node_ref) in enumerate(swarm.composite_graph.nodes.items()):
        node_indices[node_id] = index
        node_names[node_id] = node_ref.node_name

    # Iterate over the connections and probabilities
    for conn, prob in zip(swarm.connection_dist.potential_connections, edge_probs):
        src_id, dst_id = conn
        src_index = node_indices[src_id]
        dst_index = node_indices[dst_id]
        adjacency_matrix[src_index, dst_index] = prob.item()

    return adjacency_matrix, node_indices, node_names


In [6]:
def _print_conns(edge_probs: torch.Tensor, swarm):
    assert swarm is not None
    msgs = []
    for i_conn, (conn, prob) in enumerate(zip(
            swarm.connection_dist.potential_connections, edge_probs)):
        src_id, dst_id = conn
        src_node = swarm.composite_graph.find_node(src_id)
        dst_node = swarm.composite_graph.find_node(dst_id)
        src_node_name =  src_node.model_name if hasattr(src_node,"model_name") else src_node.node_name
        dst_node_name = dst_node.model_name if hasattr(dst_node,"model_name") else dst_node.node_name#
        msg = (f"{i_conn}: src={src_node_name}({src_node.id}), "
                f"dst={dst_node_name}({dst_node.id}), prob={prob.item():.3f}")
        msgs.append(msg+"\n")
        print(msg)

In [ ]:
math = {'question': ['Let A be the set of all ordered pairs of integers (m, n) such that 7m + 12n = 22. What is the greatest negative number in the set B = {m + n : (m, n) \in A}?'],
'A': ['-5'],
'B': ['-4'],
'C': ['-3'],
'D': ['-2'],
'Solution': ['B']
}
logic = {
    'question': ['Construct a complete truth table for the following pairs of propositions. Then, using the truth tables, determine whether the statements are logically equivalent or contradictory. If neither, determine whether they are consistent or inconsistent. Justify your answers. E ⊃ (F · E) and ~E · F'],
    'A': ['Logically equivalent'],
    'B': ['Contradictory'],
    'C': ['Neither logically equivalent nor contradictory, but consistent'],
    'D': ['Inconsistent'],
    'Solution': ['C']
}

In [7]:
import pandas as pd
import copy
import torch

dataset_val = CMMLUDataset('test')
print("Total df length:", len(dataset_val._total_df))
#dataset_val._total_df = dataset_val._total_df[:3000]
all_probs = []
swarm.connection_dist.model.eval()

# Initialize list to store all edge_probs
all_edge_probs = []
swarm.connection_dist.model.eval()

for i, record in enumerate(dataset_val):
    record = pd.DataFrame([record])
    input_dict = dataset_val.record_to_swarm_input(record)

    with torch.no_grad():
        edge_probs = swarm.connection_dist.get_edge_probs(swarm.composite_graph, inputs=input_dict)
    
    # Move edge_probs to CPU memory
    edge_probs = edge_probs.cpu()
    
    all_edge_probs.append(edge_probs)


# Stack all edge_probs into a tensor
all_edge_probs = torch.stack(all_edge_probs)

# Calculate mean edge_probs
mean_edge_probs = torch.mean(all_edge_probs, dim=0)
'''
# Initialize list to store distances and corresponding records
distances = []

for record in dataset_val:
    record_copy = copy.deepcopy(record)
    record_df = pd.DataFrame([record_copy])
    input_dict = dataset_val.record_to_swarm_input(record_df)
    edge_probs = swarm.connection_dist.get_edge_probs(swarm.composite_graph, inputs=input_dict)
    
    # Calculate distance to the mean
    distance = torch.norm(edge_probs - mean_edge_probs)
    
    # Store distance and record
    distances.append((distance, record_copy, edge_probs))
# Sort distances in descending order and keep top 10
top_10_records = sorted(distances, key=lambda x: x[0], reverse=True)[:10]

# Extract records from tuples
top_10_records = [(record, edge_probs) for _, record, edge_probs  in top_10_records]
print("top10:")
for record,probs in top_10_records:
    print("Question: ", record)
    _print_conns(probs, swarm)
    print("Subtracting the mean: ")
    _print_conns(mean_edge_probs-probs ,swarm)
    print("----NExt-----")
    '''

Number of topics:  67
Total number of questions:  11649
Total df length: 11649


RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:7 and cuda:1! (when checking argument for argument index in method wrapper_CUDA__index_select)

In [ ]:
for record,probs in top_10_records:
    print("Question: ", record)
    _print_conns(probs, swarm)
    print("Subtracting the mean: ")
    _print_conns(mean_edge_probs-probs ,swarm)
    print("----NExt-----")

In [12]:
print(all_edge_probs[:][0])
all_probs_variance = torch.std(all_edge_probs, dim=0)
print(all_probs_variance)
all_probs_mean = torch.mean(all_edge_probs, dim=0)
print(all_probs_variance.shape)
print(all_probs_mean.shape)
print("Mean: ")
_print_conns(all_probs_mean, swarm)
print("Variance: ")
_print_conns(all_probs_variance, swarm)

tensor([0.0016, 0.8864, 0.2510, 0.0700, 0.0464, 0.3578, 0.7493, 0.3007, 0.7684,
        0.5983, 0.9801, 0.2579, 0.9183, 0.3974, 0.2133, 0.2039, 0.4744, 0.1440,
        0.0115, 0.7529, 0.2006, 0.8509, 0.0492, 0.9855, 0.1580, 0.8205, 0.9010,
        0.3289, 0.1866, 0.9556, 0.8487, 0.0757, 0.7912, 0.7321, 0.9856, 0.9899,
        0.7145, 0.7786, 0.0045, 0.8227, 0.2271, 0.8237, 0.1396, 0.1561, 0.2381,
        0.0278, 0.9380, 0.2694, 0.9408, 0.3368, 0.9891, 0.9213, 0.1622, 0.4483,
        0.0269, 0.0035, 0.8168, 0.5094, 0.6646, 0.5575, 0.4492, 0.9254, 0.0391,
        0.8257])
tensor([0.0005, 0.0281, 0.0504, 0.0094, 0.0102, 0.0504, 0.0414, 0.0423, 0.0465,
        0.0516, 0.0025, 0.0335, 0.0203, 0.0546, 0.0378, 0.0256, 0.0306, 0.0223,
        0.0020, 0.0359, 0.0312, 0.0304, 0.0109, 0.0031, 0.0241, 0.0489, 0.0145,
        0.0336, 0.0346, 0.0088, 0.0274, 0.0080, 0.0224, 0.0300, 0.0040, 0.0033,
        0.0374, 0.0341, 0.0013, 0.0299, 0.0422, 0.0282, 0.0198, 0.0425, 0.0389,
        0.0056, 0.0135,

Mean: 
0: src=DirectAnswer(5Q3h), dst=CoTStep(6RLA), prob=0.997
1: src=DirectAnswer(5Q3h), dst=CoTStep(6Y3D), prob=0.001
2: src=DirectAnswer(5Q3h), dst=CoTStep(8y3C), prob=0.646
3: src=CoTStep(6RLA), dst=DirectAnswer(5Q3h), prob=0.001
4: src=CoTStep(6Y3D), dst=DirectAnswer(5Q3h), prob=0.001
5: src=CoTStep(8y3C), dst=DirectAnswer(5Q3h), prob=0.000
6: src=DirectAnswer(5Q3h), dst=FinalDecision(7fFJ), prob=0.988
7: src=CoTStep(6RLA), dst=FinalDecision(7fFJ), prob=0.169
8: src=CoTStep(6Y3D), dst=FinalDecision(7fFJ), prob=0.004
9: src=CoTStep(8y3C), dst=FinalDecision(7fFJ), prob=0.918
Variance: 
0: src=DirectAnswer(5Q3h), dst=CoTStep(6RLA), prob=0.000
1: src=DirectAnswer(5Q3h), dst=CoTStep(6Y3D), prob=0.000
2: src=DirectAnswer(5Q3h), dst=CoTStep(8y3C), prob=0.000
3: src=CoTStep(6RLA), dst=DirectAnswer(5Q3h), prob=0.000
4: src=CoTStep(6Y3D), dst=DirectAnswer(5Q3h), prob=0.000
5: src=CoTStep(8y3C), dst=DirectAnswer(5Q3h), prob=0.000
6: src=DirectAnswer(5Q3h), dst=FinalDecision(7fFJ), prob=0.000
7: src=CoTStep(6RLA), dst=FinalDecision(7fFJ), prob=0.000
8: src=CoTStep(6Y3D), dst=FinalDecision(7fFJ), prob=0.000
9: src=CoTStep(8y3C), dst=FinalDecision(7fFJ), prob=0.000

In [8]:
edge_probs = torch.sigmoid(swarm.connection_dist.edge_logits)
_print_conns(edge_probs, swarm)
#edge_probs = all_probs_mean
#reorder the input
#edge_probs = [*edge_probs[:-8], *edge_probs[-8::2], *edge_probs[-7::2]]
adjacency_matrix,node_indices,names = create_adjacency_matrix(edge_probs, swarm)
#reorder the matrix
print((adjacency_matrix[1::2,:]).shape)
adjacency_matrix = np.concatenate([adjacency_matrix[0,:][np.newaxis,:], adjacency_matrix[1::2,:], adjacency_matrix[2::2,:]], 0)
adjacency_matrix = np.concatenate([adjacency_matrix[:,0][:,np.newaxis], adjacency_matrix[:,1::2], adjacency_matrix[:,2::2]], 1)
_print_conns(edge_probs, swarm)
# Plot the heatmap with a smoother colormap
fig, ax = plt.subplots()
heatmap = ax.imshow(adjacency_matrix, cmap='viridis')

# Add a colorbar
cbar = plt.colorbar(heatmap)

# Set the tick labels
num_nodes = len(node_indices)
ax.set_xticks(np.arange(num_nodes))
ax.set_yticks(np.arange(num_nodes))
ax.set_xticklabels([str(node_id) for node_id in node_indices.values()])
ax.set_yticklabels([str(node_id) for node_id in node_indices.values()])

# Rotate the tick labels if needed
plt.setp(ax.get_xticklabels(), rotation=90, ha="right")

# Set the axis labels
ax.set_xlabel('Destination Node')
ax.set_ylabel('Source Node')

# Set the title
ax.set_title('Probability Heatmap')

# Show the plot
plt.show()
for node in node_indices:
    print(node_indices[node], names[node])

AttributeError: 'EdgeWiseDistributionByModel' object has no attribute 'edge_logits'

In [ ]:
adjacency_matrix = np.stack([adjacency_matrix[0::2,:], adjacency_matrix[1::2,:]])
adjacency_matrix = np.stack([adjacency_matrix[:,0::2], adjacency_matrix[:,1::2]])

In [ ]:
adjacency_matrix = np.stack([adjacency_matrix[0::2,:], adjacency_matrix[1::2,:]])
adjacency_matrix = np.stack([adjacency_matrix[:,0::2], adjacency_matrix[:,1::2]])

In [ ]:
adjacency_matrix = np.stack([adjacency_matrix[0::2,:], adjacency_matrix[1::2,:]])
adjacency_matrix = np.stack([adjacency_matrix[:,0::2], adjacency_matrix[:,1::2]])

In [ ]:
print(swarm.connection_dist.model.linear.bias)

In [ ]:
llm = LLM(model="google/gemma-7B-it", dtype="half", max_model_len=5888)

tokenizer = AutoTokenizer.from_pretrained("google/gemma-7B-it")

In [ ]:
message = [{"role": "user", "content": "Write a poem about cats."}]
prompt = tokenizer.apply_chat_template(message, tokenize=False, add_generation_prompt=True)

sampling_params = SamplingParams(temperature=0.8, top_p=0.95, top_k=50,max_tokens=2000)
outputs = llm.generate(prompt,sampling_params=sampling_params)
# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")

In [ ]:
# pip install accelerate

tokenizer = AutoTokenizer.from_pretrained("google/gemma-7B-it")
#model = AutoModelForCausalLM.from_pretrained("google/gemma-7B-it",torch_dtype=torch.bfloat16).to("cuda")
model = AutoModelForCausalLM.from_pretrained("google/gemma-7B-it", torch_dtype=torch.bfloat16).to("cuda")



In [ ]:
#print(tokenizer.special_tokens_map)
#dummy message
message = [{"role": "user", "content": "Write a poem about cats."}]
prompt = tokenizer.apply_chat_template(message, tokenize=False, add_generation_prompt=True)
prompt = tokenizer.encode(prompt, add_special_tokens=False, return_tensors="pt").to("cuda")
prompt_len = len(prompt[0])
outputs = model.generate(
            prompt,
            do_sample=True,
            max_length=2000,
            top_k=50,
            top_p=1.0
        )
output_text = tokenizer.decode(outputs[0][prompt.shape[-1]:],skip_special_tokens=True)

In [ ]:
print(output_text)

In [ ]:
# Create an instance of the GPTChat class
gpt_chat = CustomLLM()
#gpt_chat_2 = CustomLLM()

In [ ]:

#if gpt_chat is gpt_chat_2:
#    print("Same instance")
# Create a list of Message objects
messages = [Message(role="user", content="What would a dog say if he could speak?")]
# Move messages to GPU
# Create tasks for the gen method
tasks = [asyncio.create_task(gpt_chat.agen(messages))]#, asyncio.create_task(gpt_chat_2.agen(messages))]

# Wait for the tasks to complete and get the results
results = [await task for task in tasks]

# Print the output
print(results)
#print(results[1])

In [ ]:
potential_connections = [0 for _ in range(188)]
init_connection_probability = 0.1
domain = "crosswords"
llm_backbone_name="gpt2"

In [ ]:
swarm = Swarm(["CrosswordsReflection", "CrosswordsToT", "CrosswordsBruteForceOpt"], "crosswords", "gpt-3.5-turbo-1106", #"gpt-4-1106-preview"
            final_node_class="ReturnAll", 
            final_node_kwargs={},
            edge_optimize=True,
            init_connection_probability=init_connection_probability, 
            connect_output_nodes_to_final_node=connect_output_nodes_to_final_node, 
            include_inner_agent_connections=include_inner_agent_connections,
            edge_network_enable=edge_network_enable,
            llm_backbone_name=llm_backbone_name)

In [ ]:
edge_network = EdgeNetwork(llm_backbone_name=llm_backbone_name, num_edges=len(potential_connections), initial_probability=init_connection_probability)
connection_dist = EdgeWiseDistributionByModel(potential_connections, edge_network, domain)

connection_dist.load_state_dict(torch.load("result/crosswords/experiment_edge_logits_10.pt"))

In [ ]:
# Create an instance of EdgeNetwork
num_edges = 188  # Replace with the number of edges used when saving the model
llm_backbone_name = 'gpt2'  # Replace with the name of the model used when saving the model
model = EdgeNetwork(llm_backbone_name, num_edges)

# Load the state dictionary
state_dict = torch.load('result/crosswords/experiment_edge_logits_10.pt')

# Remove 'model.' prefix from state dictionary keys and exclude 'order_params'
state_dict = {k.replace('model.', ''): v for k, v in state_dict.items() if k != 'model.order_params'}

model.load_state_dict(state_dict)

In [ ]:
# Create an instance of EdgeNetwork
num_edges = 188  # Replace with the number of edges used when saving the model
llm_backbone_name = 'gpt2'  # Replace with the name of the model used when saving the model
model = EdgeNetwork(llm_backbone_name, num_edges)

# Load the state dictionary
state_dict = torch.load('result/crosswords/experiment_edge_logits_10.pt')

# Remove 'model.' prefix from state dictionary keys
state_dict = {k.replace('model.', ''): v for k, v in state_dict.items()}

model.load_state_dict(state_dict)